File: Problems/BinaryTreeRightSideView.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
   public:
    // bfs solution using a queue
    vector<int> rightSideView(TreeNode *root) {
        // using bfs and label height in each nodes and keep updating the res vector
        vector<int> res;
        if (root == nullptr)
            return res;
        queue<pair<int, TreeNode *>> q;
        pair<int, TreeNode *> p(0, root);
        q.push(p);
        while (!q.empty()) {
            pair<int, TreeNode *> pp = q.front();
            q.pop();
            int h = pp.first;
            TreeNode *node = pp.second;
            if (res.size() > h)
                res[h] = node->val;
            else
                res.push_back(node->val);
            if (node->left != nullptr)
                q.push(pair<int, TreeNode *>(h + 1, node->left));
            if (node->right != nullptr)
                q.push(pair<int, TreeNode *>(h + 1, node->right));
        }
        return res;
    }

    // recursive solution usign preorder traversal
    void recursion(TreeNode *root, int level, vector<int> &res) {
        if (root == NULL) return;
        if (res.size() < level) res.push_back(root->val);
        recursion(root->right, level + 1, res);
        recursion(root->left, level + 1, res);
    }

    vector<int> rightSideView2(TreeNode *root) {
        vector<int> res;
        recursion(root, 1, res);
        return res;
    }
};

File: Problems/BinaryTreeRightSideView.py
# iterative solution
# it is like bfs; traverse level by level and put nodes into a list
# and choose the last node, which must be the right most node in each height of tree
def rightSideView(self, root):
    view = []
    if root:
        level = [root]
        while level:
            view += level[-1].val,
            level = [kid for node in level for kid in (
                node.left, node.right) if kid]
    return view


# recursive solution
# modified preorder traversal
# in each height of tree, it always visits the node at the most right; then append the value into a list
# only the right most node in each height can satisfy the condition (h == len(view))
def rightSideView(self, root):
    def dfs(node, h):
        if node:
            if h == len(view):
                view.append(node.val)
            dfs(node.right, h + 1)
            dfs(node.left, h + 1)
    view = []
    dfs(root, 0)
    return view
